gateway/src/main/java/io/zeebe/gateway/grpc/GrpcErrorMapper.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.0. You may not use this file
+ * except in compliance with the Zeebe Community License 1.0.
+ */
+package io.zeebe.gateway.grpc;
+
+import com.google.rpc.Code;
+import com.google.rpc.Status;
+import com.google.rpc.Status.Builder;
+import io.grpc.StatusRuntimeException;
+import io.grpc.protobuf.StatusProto;
+import io.zeebe.gateway.Loggers;
+import io.zeebe.gateway.cmd.BrokerErrorException;
+import io.zeebe.gateway.cmd.BrokerRejectionException;
+import io.zeebe.gateway.cmd.InvalidBrokerRequestArgumentException;
+import io.zeebe.gateway.cmd.PartitionNotFoundException;
+import io.zeebe.gateway.impl.broker.response.BrokerError;
+import io.zeebe.gateway.impl.broker.response.BrokerRejection;
+import io.zeebe.msgpack.MsgpackPropertyException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeoutException;
+import org.slf4j.Logger;
+
+/** Maps arbitrary {@link Throwable} to {@link StatusRuntimeException} and logs the exception. */
+public final class GrpcErrorMapper {
+  private final Logger logger;
+
+  public GrpcErrorMapper() {
+    this(Loggers.GATEWAY_LOGGER);
+  }
+
+  public GrpcErrorMapper(final Logger logger) {
+    this.logger = logger;
+  }
+
+  public StatusRuntimeException mapError(final Throwable error) {
+    return StatusProto.toStatusRuntimeException(mapErrorToStatus(error));
+  }
+
+  private Status mapErrorToStatus(final Throwable error) {
+    final Builder builder = Status.newBuilder();
+
+    if (error instanceof ExecutionException) {
+      return mapErrorToStatus(error.getCause());
+    } else if (error instanceof BrokerErrorException) {
+      final Status status = mapBrokerErrorToStatus(((BrokerErrorException) error).getError());
+      builder.mergeFrom(status);
+
+      // When there is back pressure, there will be a lot of `RESOURCE_EXHAUSTED` errors and the log
+      // can get flooded, so log them at the lowest level possible
+      if (status.getCode() != Code.RESOURCE_EXHAUSTED.getNumber()) {
+        logger.error("Expected to handle gRPC request, but received error from broker", error);
```

Comment:
Currently thinking whether it really necessary to error in gateway in general, since they properly logged in the broker already and when the client receives the error

Reply:
I'm not sure we log all errors in the broker. For example, partition mismatch i think is not logged, maybe others?

Reply:
But is this really an error then is the question? It is kind of expected that this can happen - ideally we should just retry that in this case

Reply:
My rule of thumb in general is, assuming a client-server setup, if the server returns an error, the client should log it, not the server. However, the server might log error details that are not returned to the client - i.e. sensitive info, information that isn't helpful for the client, etc. 

In this particular case, we return an INTERNAL error to the client (e.g. 500 - which, by the way, is probably the wrong code to return for `PARTITION_LEADER_MISMATCH` :sweat_smile:), with very little information from the client point of view. So we need to log it in the broker/gateway. From the broker point of view though, you're right, this is an expected error - hence why I'd log it on the "client" side, i.e. the gateway in this instance. But like I said, bad example, as `PARTITION_LEADER_MISMATCH` should probably return `UNAVAILABLE` or something like this, indicating the user can retry the request :sweat_smile: